src/github.com/matrix-org/dendrite/roomserver/types/types.go

```diff
@@ -1,10 +1,82 @@
 // Package types provides the types that are used internally within the roomserver.
 package types
 
+import (
+	"github.com/matrix-org/gomatrixserverlib"
+)
+
 // A PartitionOffset is the offset into a partition of the input log.
 type PartitionOffset struct {
 	// The ID of the partition.
 	Partition int32
 	// The offset into the partition.
 	Offset int64
 }
+
+// A StateKey is a pair of a numeric event type and a numeric state key.
+// It is used to lookup state entries.
+type StateKey struct {
+	// The numeric ID for the event type.
+	EventTypeNID int64
+	// The numeric ID for the state key or 0 if the event is not a state event.
+	EventStateKeyNID int64
+}
+
+// LessThan returns true if this state key is less than the other state key.
+func (a StateKey) LessThan(b StateKey) bool {
+	if a.EventTypeNID != b.EventTypeNID {
+		return a.EventTypeNID < b.EventTypeNID
+	}
+	return a.EventStateKeyNID < b.EventStateKeyNID
+}
+
+// A StateEntry is an entry in the room state of a matrix room.
+type StateEntry struct {
+	StateKey
+	// The numeric ID for the event.
+	EventNID int64
+}
+
+// LessThan returns true if this state entry is less than the other state entry.
+func (a StateEntry) LessThan(b StateEntry) bool {
```

Comment:
I think you should add a comment here to explain how you plan on using sorting to detect and remove duplicates. The `LessThan` implementation exists only for this reason, and that there is no semantic meaning to the actual sort output.